### ISSUE TITLE
Refresh a node using a timer?

### ISSUE THREAD

**Issue by ZodiacFRA** [JacquesLucke/animation_nodes#985]
**Hello and happy Holidays everybody :)**

I'm currently trying to achieve real-time audio visualisation in blender, and at the moment it's working great with a `ModalTimerOperator`. 
However I would like to do the same thing using custom node trees, which is a much more user friendly way of doing what I want. 
I searched everywhere for a template of a `timer refreshed node` but nothing came out and I was wondering if you could give me some leads on this :)

Thanks and have a nice day!

**Comment by OmarSquircleArt** [comment]
Nodes in AN can't execute on their own, they form a node tree and this node tree is then executed. If you are looking to make the node tree execute every `t` second, then have a look at the [*Minimum Time Difference*](https://animation-nodes-manual.readthedocs.io/en/latest/user_guide/getting_started/execution.html#minimum-time-difference) option coupled with the *Always* auto execution option.

**Comment by OmarSquircleArt** [comment]
@ZodiacFRA I see. I don't fully understand the auto execution implementation in AN so I may not be able to help you. However, a new timer API was recently added that may be what you are looking for, see the [*Run a function every x second example*](https://docs.blender.org/api/blender2.8/bpy.app.timers.html#run-a-function-n-times-every-x-seconds). Is this what you are looking for?

**Comment by Clockmender** [comment]
@ZodiacFRA Can you explain how you are getting the mic values into Blender?

Auto-execution can run in various modes, unchecking this means that AN does not execute if you render the animation, or run the Timeline:
_Always_ - runs all the time with no control, I use this to test a node tree without running the animation, I never use this to render an animation as the node tree may execute multiple times between rendered frames.
_Frame Changed_ - runs when you advance the Timeline.
_Property Changed_ - a node input, or option changes for example.
_Trigger_ - A referenced object moves, or rotates for example, you can have multiple triggers BTW.
_Minimum Time_ - can be used to make sure AN only executes at a set time interval, say every second (this would equate to every 24 frames if you set the project frame rate to 24fps, each frame is then 41.67 milliseconds long).

The reference @OmarSquircleArt quoted above would run a Blender script every programmed time interval and  is not part of AN, just for ordinary scripts, as I see it. I have not used this, I do all my scripting in AN using script nodes and execution by one, or more methods above.

If you are using a Blender script to get the mic values from a live mic and then use these, can you not set its execution to run every so many seconds, or part thereof, using the reference made above and then use the same script to change the location of an object?

There exists a Sound Bake Node to use a recorded sound to drive an animation, I am guessing this is of no use as you are wanting live input from the mic to animate an object. I don't understand what the end goal is, as this method cannot be used to make a rendered, animated video.

Perhaps you could share your project, its goals and exactly what you want to achieve with us. I feel we are having to guess a little in order to help you, hence why I have not replied before.

Cheers, Clock.

**Comment by ZodiacFRA** [comment]
@Clockmender thanks for the detailed reply, As i'm at the very beginning of my project it's not yet online, but I plan publishing it under GNU as soon as I'll have a solid starting point.

I'm trying to create parametric realtime animation driven by sound and MIDI inputs.

I'll use Eevee fullscreen so the result will pretty close to a full render. 
_If the 2.8 API is not ready for this I'll use the classic viewport and modify the video with another real time video software to enhance it_

As you guessed I'm not interested by sound bake Node because I wanna use this for performance purposes. 

Atm i'm using pyAudio with ASIO to get my 4 input channels values into blender, and pygame for the MIDI input part.
Each time the Modal operator timer is triggered I read the buffers and transform the values into comprehensible ones (Audio analysis, slices etc)

This is working as I want using a TimerModalOperator, but exposing the values via a node tree (and choosing graphically what they affect) is much more user friendly than hardcoding the input-output links.

I never heard about auto-executing multiple times the same script in Blender, do you have a link or something? 
I also thought about a script with a timer, calling the node tree update but I do not know how to do this as well :)
Have a nice day!

**Comment by Clockmender** [comment]
> I'm trying to create parametric realtime animation driven by sound and MIDI inputs.

I have done quite a lot of work with MIDI also, see my website [here](https://clockmender.uk/blender/animation-nodes/) and [here](https://clockmender.uk/blender/animation-nodes/midi-controls/) for details of this work, I think it  could be adapted to run real-time. I will write more later, after I complete Mrs. Clockmender's list...

Cheers, Clock.

**Comment by Clockmender** [comment]
> I also thought about a script with a timer, calling the node tree update but I do not know how to do this as well

Like in the picture and blend file below! Basically you can set Auto Execution => Always then set the Minimum Time Difference to make it execute at some time extent, I,e, 0.04 = 40 milliseconds, or once every 25th of a second.

This may also help you, I am not sure where you are with getting MIDI into Blender:

![screen shot 2018-12-28 at 20 32 43](https://user-images.githubusercontent.com/29657804/50527378-c45e2600-0adf-11e9-914f-492262b5fd7f.png)

This project just waits for an input into the script node, denoted "Note_i", for now I am feeding this from a list of real MIDI data, just four examples. Changing the index number in the Get List Element Node triggers the node tree to execute, which will rotate the appropriate piano key as detailed in the MIDI input string (just cycle from 0 to 3).  The node tree will only execute every 0.1 second BTW as I have set it up. If you can feed this with MIDI data from your system, the notes should press down when my script gets a "note_on" event and return to normal when it gets a "note_off" event.

Feeding it from a flat file of multiple MIDI events would entail adding a loop to read the file, something like this:

```
with open(midi_file) as f1:
                for line in f1:
                    in_l = [elt.strip() for elt in line.split(',')]
                    if (in_l[2] == 'Header'):
                        # Get Pulse variable.
                        pulse = int(in_l[5])

                    elif (in_l[2] == 'Tempo'):
                        if (in_l[0] == '1'):
                            # Get Initial Tempo.
                            tempo = in_l[3]
                            self.bpm = float( round( (60000000 / int(tempo)), 5) )
                            bps = float( round( (self.bpm / 60), 5) )
                        else:
                            # Add Tempo Changes to events list.
                            events_list.append(in_l)

                        etc. etc....
```

What I don't know is how the data is provided from your systems, i.e. is it ASCII, is it in standard MIDI format, does it have a time sig as the input strings I have used have, all of this might result in changes to the code, but I can help here with this.

Here is the blend file (has to be zipped to post here): [real-time-midi.blend.zip](https://github.com/JacquesLucke/animation_nodes/files/2715650/real-time-midi.blend.zip)

You will need official release 2.79 Blender and official release 2.0 Animation Nodes for things to work.

Here is an example from a real MIDI .csv file:

```
0, 0, Header, 1, 11, 15360
1, 0, Start_track
1, 0, Title_t, "Transport"
1, 0, Instrument_name_t, "Hardware Interface II"
1, 0, MIDI_port, 0
1, 0, Tempo, 731707
1, 0, Time_signature, 4, 2, 24, 8
1, 4055040, End_track
2, 0, Start_track
2, 0, Title_t, "Fretless Bass Ch2"
2, 0, Instrument_name_t, "Fretless Bass Ch2"
2, 0, MIDI_port, 13
2, 0, Control_c, 0, 64, 0
2, 122880, Note_on_c, 0, 38, 79
2, 142464, Note_off_c, 0, 38, 0
2, 145920, Note_on_c, 0, 38, 85
2, 148736, Note_off_c, 0, 38, 0
2, 153600, Note_on_c, 0, 38, 75
2, 176640, Note_off_c, 0, 38, 0
2, 176640, Note_on_c, 0, 45, 72
2, 183808, Note_off_c, 0, 45, 0
```

Let me know if this helps you.

Cheers, Clock.

**Comment by Clockmender** [comment]
Hi @ZodiacFRA 

> Atm i'm using pyAudio with ASIO to get my 4 input channels values into blender, and pygame for the MIDI input part.

How are you doing this, I have a StudioLogic TMK88 keyboard with iRig interface (Currently on my iMac connected to Reason DAW), so can test this if you can point me in the right direction to get midi input from my keyboard.

Cheers, Clock.

**Comment by ZodiacFRA** [comment]
@Clockmender Thanks for all those replies
I'm on holidays so I don't have all the files ATM sorry, I'll upload them for you when I'll come home :)

My current refresh rate (with ModalTimerOperator) is set at 0.0001sec. There is no reason to this particular value, but this gives really low latency and it's not destroying the CPU (< 5% change from what I saw)

The MIDI I get from the MatrixBrute and MS2000 is standard protocol, something like
`[Note ON/OFF, [Note, Velocity, ?] Timestamp]`, really easy to parse
I haven't tried sending parameters values yet.
The Audio I get is stored in numpy arrays, and I detect the peak volume of the small buffer I just read.

I'm currently with my laptop, which uses python 3.7.1, and it's not compatible with AN (I should not have upgraded to Fedora 27, Python 3.7 is a retro compatibility nightmare.)
I've tried to build AN myself but found no documentation (https://animation-nodes-manual.readthedocs.io/en/latest/dev_guide/setup/linux.html)
so I won't be able to test your file until I come home :(

If I understand correctly, you AN setup read 4 text files containing CSV MIDI data, which are parsed and sent to an AN output node?
If I use the same method, I'll need to have a script executed only once, initializing pyAudio and pyGame, and a script refreshed everytime, reading the buffers.

I'm not really familiar with AN:
-Is the `Invoke subprogram` node the one used to run the script shown on the bottom right?
-Can you choose the number of output sockets of this node?
-With Auto-refresh on, are all the script nodes refreshed? (So I can have a one script node for midi, one for audio, and don't reinitialize libraries each time)

Happy Holidays!

**Comment by Clockmender** [comment]
This can wait until you get back home, although I am not very patient at times when I get started on something!

> If I understand correctly, you AN setup read 4 text files containing CSV MIDI data, which are parsed and sent to an AN output node?

Not quite, I am just reading four individual lines from one MIDI file in this case, just to test it.

> -Is the Invoke subprogram node the one used to run the script shown on the bottom right?

Yes!

> -Can you set the number and type of output sockets of this node?

Yes! All input/output sockets are programmable in the script node structure, my aim is to develop my own Realtime MIDI node to go with the ones I have already written to read MIDI CSV files to animate objects. We could work together on this to use Realtime Audio as well.

-With Auto-refresh on, are all the script nodes refreshed?

Yes! You can set when the scripts are executed, per frame, always, time intervals, etc. I think a "Time Interval Limited", "Always" set up would work very well. Something like 0.002 would execute every 2 milliseconds, in line with my Linux box, which uses an E-Mu 0404 interface seems quite fast enough for realtime playing MIDI instruments.

I just need some help getting started with pygame, for example, or whatever you use, so I can set a variable to the input from my keyboard through my iRig Pro USB interface on my Mac.

Cheers, Clock.

**Comment by Clockmender** [comment]
Sorry Mate, I have tied very thing I know and whilst I can get pygame to work with Python on my machine:

![screen shot 2018-12-31 at 16 35 45](https://user-images.githubusercontent.com/29657804/50575288-31313300-0df4-11e9-8abb-63eee5cb6bbe.png)

I cannot get it to load as a module into Blender, have you achieved this? I have spent two days on this and got exactly nowhere, it just won't add into the Blender Python Site-packages at all. I even tried copying the one from my Python install, but got a message saying pygame.base not found when I tried to "import pygame.midi" in the script node.

Cheers, Clock.

**Comment by Clockmender** [comment]
Eureka! I hove found it...

So I got really fed up and tried my last resort, i.e. went to my flying club and sat in the corner undisturbed by wifey, etc. and did this:

1) Install pip into the Blender Python directory on my Mac - using [this](https://blender.stackexchange.com/questions/56011/how-to-install-pip-for-blenders-bundled-python/56013#56013).

2) Get the pygame .whl file from [here](https://bitbucket.org/pygame/pygame/downloads/). I chose the python 3.5 version for macOS.

3) Navigate to the Blender.app python bin file and install it like this in a terminal window:

`./pip install /Users/mac/Downloads/pygame-1.9.4-cp35-cp35m-win_amd64.whl`

Note the _./pip_ bit so it uses the **local** pip not the system wide pip...

Now I have got this far, with my iRig connected, but no MIDI keyboard:

![screen shot 2019-01-02 at 10 34 55](https://user-images.githubusercontent.com/29657804/50588476-1acdba80-0e7a-11e9-91f6-febaa0a4a33c.png)

And when I get home, assuming my wings don't fall off, I will connect my keyboard to my iRig, ad some code to read the MIDI info into variables and see what happens...

Cheers, Clock.

**Comment by ZodiacFRA** [comment]
Hello!
I'm sorry I did not see your messages till today, **I went through the exact same problems on my windows 10 machine**, and I had to manually add pip to blender's python and add the libraries using this pip like you did (but my Linux blender uses system python).
I sent you a message from your site's contact form so you have my email address.
I'll be home tonight so I'll send you my code, did you try to install pyAudio as well? It needs portaudio as a dependancy.

**Comment by TesseraSkye** [comment]
@ZodiacFRA 
Hey, sorry for the ping, but uh.. Looks like we're trying to do the exact same thing with real-time vis. I'd be interested in passing notes, if you're up for it.

**Comment by Clockmender** [comment]
@TesseraSkye - Hi

Got your message here, @ZodiacFRA and I have got as far as getting real time MIDI working well in Animation Nodes and are currently working on real-time audio. @ZodiacFRA has a working node for inputting audio, but I am having issues loading the pyaudio python module on my Mac just now, hope to have that resolved shortly. If you are interested in swapping notes, le me know here, or mail via my contact details here. I am sure @ZodiacFRA will also be interested in talking. How far have you got to date with this?

I see you posted in the middle of our night (circa 4 AM), where in the world are you?

Cheers, Clock.

PS. here's some  images:

![img_2029](https://user-images.githubusercontent.com/29657804/51165424-0b963600-1898-11e9-85ef-63b45ccdd66b.jpg)

_Blender and Reason DAW together._

![screen shot 2019-01-13 at 20 20 13](https://user-images.githubusercontent.com/29657804/51165433-118c1700-1898-11e9-8735-9d8339fae2db.png)

_Keyboard operated digger arms._

![screen_shot_2019-01-12_at_18 35 40](https://user-images.githubusercontent.com/29657804/51165444-181a8e80-1898-11e9-9b29-5f0ccd323b41.png)

_MIDI Keys, Pedal and Mod Wheel working._

**Comment by TesseraSkye** [comment]
Hello!
The last month has been a pretty steep learning curve, so I’m not incredibly far along, but I’ve written backend for an OSC server node (or at least, learned how to use pythonosc in such a way that I can implement a node). I’m working on midi with mido, audio with sounddevice, and will be adding a few other things like art-net and SMPTE.

I’m on mountain time, I was just up late :P

**Comment by Clockmender** [comment]
We are in UK and France, so we are on GMT and a little bit!

We are trying sounddevice now for audio, having tried PyAudio, works on Windows, but there are no wheel files for my Macs, so I couldn't load the modules into Blender's python (all sorts of issues here, even SSL certificates invalid for the source files). The new sounddevice Init node is going well, but not finished yet, more details after we progress a bit. Have you managed to read the input from sounddevice into either python buffers, or numb arrays yet in a node? I did not get that far tonight before my old brain gave up for the night....

For the MIDI we used PyGame.Midi seemed a good option and we have it all working inside Animation Nodes, we are happy with this, we did not try Mido.

Cheers, Clock.
